compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -983,6 +1013,19 @@ fn overloaded_place<'a, 'tcx>(
     ExprKind::Deref { arg: ref_expr.to_ref() }
 }
 
+fn fake_read_capture_upvar<'tcx>(
+    cx: &mut Cx<'_, 'tcx>,
+    closure_expr: &'tcx hir::Expr<'tcx>,
+    hir_id: hir::HirId,
+) -> Expr<'tcx> {
+    let temp_lifetime = cx.region_scope_tree.temporary_scope(closure_expr.hir_id.local_id);
+    let var_ty = cx.typeck_results().node_type(hir_id);
+
+    // FIXME: Maybe we should introduce an immutable borrow of the fake capture so that we don't
+    // end up moving this place
+    Expr { temp_lifetime, ty: var_ty, span: closure_expr.span, kind: convert_var(cx, hir_id) }
```

Comment:
You want all of this:

https://github.com/sexxi-goose/rust/pull/45/files#diff-ad5988b23007e07f23d5459907b7aa75c9dc6ae82623be8cd4bc6acc58cb3397R1056-R1073